File: 2025/02 February/02 Level order traversal/main.py
# Link: https://www.geeksforgeeks.org/problems/level-order-traversal/1

from collections import deque


# Tree Node
class Node:

    def __init__(self, val):
        self.right = None
        self.data = val
        self.left = None


# Function to Build Tree
def buildTree(s):
    # Corner Case
    if len(s) == 0 or s[0] == "N":
        return None

    # Creating list of strings from input
    # string after spliting by space
    ip = list(map(str, s.split()))

    # Create the root of the tree
    root = Node(int(ip[0]))
    size = 0
    q = deque()

    # Push the root to the queue
    q.append(root)
    size = size + 1

    # Starting from the second element
    i = 1
    while size > 0 and i < len(ip):
        # Get and remove the front of the queue
        currNode = q[0]
        q.popleft()
        size = size - 1

        # Get the current node's value from the string
        currVal = ip[i]

        # If the left child is not null
        if currVal != "N":

            # Create the left child for the current node
            currNode.left = Node(int(currVal))

            # Push it to the queue
            q.append(currNode.left)
            size = size + 1
        # For the right child
        i = i + 1
        if i >= len(ip):
            break
        currVal = ip[i]

        # If the right child is not null
        if currVal != "N":

            # Create the right child for the current node
            currNode.right = Node(int(currVal))

            # Push it to the queue
            q.append(currNode.right)
            size = size + 1
        i = i + 1
    return root


class Solution:
    def levelOrder(self, root):
        vt = []
        if not root:
            return vt

        q = [root]
        while q:
            n = len(q)
            sol = []

            for _ in range(n):
                tmp = q.pop(0)
                if tmp.left is not None:
                    q.append(tmp.left)
                if tmp.right is not None:
                    q.append(tmp.right)
                sol.append(tmp.data)
            vt.append(sol)
        return vt


if __name__ == "__main__":
    obj = Solution()
    s = "1 2 3"
    root = buildTree(s)
    res = obj.levelOrder(root)
    for i in res:
        print(i, end=" ")
    print()

    s = "10 20 30 40 50"
    root = buildTree(s)
    res = obj.levelOrder(root)
    for i in res:
        print(i, end=" ")
    print()

    s = "1 3 2 N N N 4 6 5"
    root = buildTree(s)
    res = obj.levelOrder(root)
    for i in res:
        print(i, end=" ")
    print()

File: 2025/02 February/02 Level order traversal/main.cpp
// Link: https://www.geeksforgeeks.org/problems/level-order-traversal/1

#include <iostream>
#include <vector>
#include <queue>
#include <sstream>
using namespace std;

class Node
{
public:
    int data;
    Node *left;
    Node *right;

    // Constructor
    Node(int val)
    {
        data = val;
        left = nullptr;
        right = nullptr;
    }
};

// Utility function to create a new Tree Node
Node *newNode(int val)
{
    return new Node(val);
}

void inOrder(Node *node)
{
    if (node == nullptr)
        return;

    inOrder(node->left);
    printf("%d ", node->data);

    inOrder(node->right);
}

// Function to Build Tree
Node *buildTree(string str)
{
    // Corner Case
    if (str.length() == 0 || str[0] == 'N')
        return nullptr;

    // Creating vector of strings from input
    // string after spliting by space
    vector<string> ip;

    istringstream iss(str);
    for (string str; iss >> str;)
        ip.push_back(str);

    // Create the root of the tree
    Node *root = newNode(stoi(ip[0]));

    // Push the root to the queue
    queue<Node *> queue;
    queue.push(root);

    // Starting from the second element
    int i = 1;
    while (!queue.empty() && i < ip.size())
    {

        // Get and remove the front of the queue
        Node *currNode = queue.front();
        queue.pop();

        // Get the current node's value from the string
        string currVal = ip[i];

        // If the left child is not null
        if (currVal != "N")
        {

            // Create the left child for the current node
            currNode->left = newNode(stoi(currVal));

            // Push it to the queue
            queue.push(currNode->left);
        }

        // For the right child
        i++;
        if (i >= ip.size())
            break;
        currVal = ip[i];

        // If the right child is not null
        if (currVal != "N")
        {

            // Create the right child for the current node
            currNode->right = newNode(stoi(currVal));

            // Push it to the queue
            queue.push(currNode->right);
        }
        i++;
    }

    return root;
}

class Solution
{
public:
    vector<vector<int>> levelOrder(Node *root)
    {
        vector<vector<int>> vt;

        if (!root)
        {
            return vt;
        }

        queue<Node *> q;
        q.push(root);

        while (!q.empty())
        {
            int n = q.size();
            vector<int> sol;

            for (int i = 0; i < n; i++)
            {
                Node *tmp = q.front();
                q.pop();

                if (tmp->left != NULL)
                {
                    q.push(tmp->left);
                }
                if (tmp->right != NULL)
                {
                    q.push(tmp->right);
                }
                sol.push_back(tmp->data);
            }
            vt.push_back(sol);
        }
        return vt;
    }
};

int main()
{
    Solution object;
    string s = "1 2 3";
    Node *root = buildTree(s);
    vector<vector<int>> result = object.levelOrder(root);
    for (int i = 0; i < result.size(); i++)
    {
        for (int j = 0; j < result[i].size(); j++)
        {
            cout << result[i][j] << " ";
        }
        cout << endl;
    }
    cout << endl;

    s = "10 20 30 40 50";
    root = buildTree(s);
    result = object.levelOrder(root);
    for (int i = 0; i < result.size(); i++)
    {
        for (int j = 0; j < result[i].size(); j++)
        {
            cout << result[i][j] << " ";
        }
        cout << endl;
    }
    cout << endl;

    s = "1 3 2 N N N 4 6 5";
    root = buildTree(s);
    result = object.levelOrder(root);
    for (int i = 0; i < result.size(); i++)
    {
        for (int j = 0; j < result[i].size(); j++)
        {
            cout << result[i][j] << " ";
        }
        cout << endl;
    }
    cout << endl;

    return 0;
}